### ISSUE TITLE
GVN does not handle `readonly` properly

### ISSUE THREAD

**Issue by nhusung** [llvm/llvm-project#65719]
Consider the following function:

```llvm
define void @foo(ptr readonly %0, i32 noundef %1) unnamed_addr {
  %3 = load ptr, ptr %0, align 8
  %4 = icmp eq ptr %3, %0
  br i1 %4, label %6, label %5

5:                                                ; preds = %2
  br label %8

6:                                                ; preds = %2
  %7 = getelementptr inbounds { ptr, i32 }, ptr %3, i64 0, i32 1
  store i32 42, ptr %7, align 4
  ret void

8:                                                ; preds = %5, %8
  br label %8
}
```

GVN replaces the definition of `%7` by `%7 = getelementptr inbounds { ptr, i32 }, ptr %0, i64 0, i32 1`, so InstCombine eventually removes the subsequent store, since it is derived from a `readonly` pointer. To the best of my knowledge this is incorrect. The language reference explicitly states: “This attribute indicates that the function does not write through this pointer argument, *even though it may write to the memory that the pointer points to.*” Alive2 also complains about this: https://alive2.llvm.org/ce/z/Lw8eFH

To add a bit of context: I have a Rust project which is affected by this issue. I reduced the problem to this (ugly) code: https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=f4089f905148ea6fb4c28c0cb0c1a076 The program should print 42 (which it actually does in Debug mode and when executed by Miri) but actually prints 0 in release mode. The `foo` function from above corresponds to `Inner::write_outer()`. The issue is present in both rustc stable (1.72.0) and nightly (1.74.0, 2023-09-07), so LLVM 16 and 17 are affected.

**Comment by nhusung** [comment]
Note that `%0` always points to itself. The high-level idea is that if we know certain aspects of the `struct`’s memory layout, we should be able to obtain a reference to the outer `struct` from the inner `struct` simply using pointer arithmetic. However, Rust’s proposed aliasing models (Stacked Borrows and Tree Borrows) impose further restrictions, which I could only fulfill by storing the pointer to the outer `struct`. However, my idea was that the load of the pointer could be optimized away if I marked the case where `%0 != load %0` as unreachable. Hence, I originally used Rust’s [`unreachable_unchecked()`](https://doc.rust-lang.org/std/hint/fn.unreachable_unchecked.html), which generates an `unreachable` LLVM IR instruction. However, `unreachable_uncheked` is `unsafe`, so I tried to reproduce the issue in “safer” code.
